### PR TITLE
fix: refactor menu generation logic for React sidenav

### PR DESCRIPTION
**Pull request by swrichards**
## Summary

This ensures unpublished CMS pages do not appear in the sidenav menu. More generally, this PR involves a significant refactor of the earlier code with some additional tests covering a variety of edge cases. The logic was getting a bit hard to follow with all the applicable conditions, so I tried to break the logic up a bit into readable chunks.

## Issue References

- Taiga Issue: [#3479](https://taiga.maykinmedia.nl/project/open-inwoner/issue/3479)

## Checklist

- [x] CHANGELOG.rst updated

### CODE REVIEW

**Comment by swrichards** [comment]
@pi-sigma Still working on the tests (CMS headaches), but here's the basic idea for an early review.

File: src/open_inwoner/components/templatetags/side_navigation.py [hunk 1]
```diff
@@ -1,207 +1,194 @@
...
 
+        return self._process_nodes_to_json(target_nodes)
 
-@register.simple_tag(takes_context=True)
-def react_sidenav_data(context):
-    """Template tag to provide menu data for React SideNavModule component"""
-    request = context["request"]
+    def _extract_icon(self, node: NavigationNode) -> str | None:
+        try:
+            if node_id := getattr(node, "id", None):
```
Comment (pi-sigma) [inline src/open_inwoner/components/templatetags/side_navigation.py]: 1. The `try` bock can be flattened and made more specific:
```
if not node_id := getattr(node, "id", None):
    return None

try:
    page = Page.objects.get(pk=node_id)
    common_ext = CommonExtension.objects.get(extended_object=page)
except (Page.DoesNotExist, CommonExtension.DoesNotExist):
    logger.exception("No CMS page with commenextension found")
    return None

if icon := getattr(common_ext, "menu_icon", None):
    logger.debug(" Found icon via CommonExtension: %s", icon)
    return icon

return None
  ```

File: src/open_inwoner/components/templatetags/side_navigation.py [hunk 2]
```diff
@@ -231,6 +210,46 @@ def react_sidenav_data(context):
         )
         return complete_menu
 
+
+def get_extra_menu_items(context) -> List[MenuItem]:
+    extra_items: List[MenuItem] = []
+    request = context.get("request")
+
+    # FAQ item
+    if context.get("has_general_faq_questions", False):
```
Comment (pi-sigma) [inline src/open_inwoner/components/templatetags/side_navigation.py]: I'm not following the logic here. You `except NoReverseMatch` for `faq_url = reverse("general_faq")` but then you do `"href": reverse("general_faq")` again inside another `try/except`. If I understand what's supposed to happen (add the faq if the context has it and we can reverse the relevant URL), something like the following seems clearer:

```
if not context.get("has_general_faq_questions", False):
    return extra_items

if not request or not hasattr(request, "path"):
    return extra_items

try:
    faq_url = reverse("general_faq")
except NoReverseMatch:
    logger.warning("Could not add FAQ menu item: %s", e)
    return extra_items

is_current = request.path.startswith("/faq/")
extra_items.append(
    {
        "href": faq_url,
        "label": str(_("Veelgestelde vragen")),
        "icon": "question_answer",
        "current": is_current,
        "counter": None,
    }
)
logger.debug("Generated %s extra menu items", len(extra_items))

return extra_items
```

File: src/open_inwoner/components/templatetags/side_navigation.py [hunk 3]
```diff
@@ -1,207 +1,194 @@
...
+
+    def _is_current_page(self, node: NavigationNode, url: str) -> bool:
+        current = getattr(node, "selected", False)
+
+        # Handle special case where route doesn't match CMS page URLs
+        try:
+            resolved = resolve(self.request.path)
+            if resolved.url_name == "contactmoment_list":
+                current = "contactmomenten" in url
+        except Exception as e:
```
Comment (pi-sigma) [inline src/open_inwoner/components/templatetags/side_navigation.py:158]: Error handling can be more specific:
```
try:
    resolved = resolve(self.request.path)
except NoReverseMatch as exc:
    logger.debug("Could not resolve current path for menu highlighting: %s", exc)
else:
    if resolved.url_name == "contactmoment_list":
        current = "contactmomenten" in url

return current

File: src/open_inwoner/components/templatetags/side_navigation.py [hunk 4]
```diff
@@ -1,207 +1,194 @@
...
-        if home_node:
-            # Use children of home node (preferred)
-            target_nodes = getattr(home_node, "children", [])
-            logger.debug("Using home node children: %s items", len(target_nodes))
-        else:
-            # Fallback: use all visible nodes
+        return current
+
+    def _is_visible_to_user(self, node: NavigationNode) -> bool:
+        try:
```
Comment (pi-sigma) [inline src/open_inwoner/components/templatetags/side_navigation.py:164]: Here as elsewhere there's a lot of stuff in the `try` block, making it somewhat difficult to single out the potential exceptions. What is the `AttributeError` excepted for? Every CMS page should have the `is_published` method and the `publisher_is_draft` attribute, so these shouldn't fail (did you encounter an exception here while testing manually? I might be wrong here). Or could we be dealing with a missing request? Otherwise `Page.objects.get(pk=node.id)` seems to be the only thing that could cause an exception.

File: src/open_inwoner/components/templatetags/side_navigation.py [hunk 5]
```diff
@@ -1,207 +1,194 @@
...
-                logger.debug("Found icon via node.menu_icon: %s", icon)
-            elif hasattr(node, "attr") and hasattr(node.attr, "menu_icon"):
-                icon = node.attr.menu_icon
-                logger.debug("Found icon via node.attr.menu_icon: %s", icon)
-            else:
-                # Try to get it from the page's CommonExtension
-                try:
-                    if hasattr(node, "id") and node.id:
-                        from cms.models import Page
+        # Check if page is published
```
Comment (pi-sigma) [inline src/open_inwoner/components/templatetags/side_navigation.py]: `_is_visible_to_user` checks for more than just publication status, so the comment is incomplete; make more specific or remove

**Review by pi-sigma — CHANGES_REQUESTED** [review]

File: src/open_inwoner/components/templatetags/side_navigation.py [hunk 6]
```diff
@@ -231,8 +279,48 @@ def react_sidenav_data(context):
...
+        try:
+            faq_url = reverse("general_faq")
+            is_current = request.path.startswith(faq_url)
+        except NoReverseMatch:
+            logger.warning("Could not add FAQ menu item: %s", exc_info=True)
+            return extra_items
+
+        extra_items.append(
+            {
+                "href": reverse("general_faq"),
```
Comment (pi-sigma) [inline src/open_inwoner/components/templatetags/side_navigation.py]:
```suggestion
                "href": faq_url,
```

File: src/open_inwoner/components/templatetags/side_navigation.py [hunk 7]
```diff
@@ -1,207 +1,263 @@
...
 
-            # Skip hidden items
-            if not getattr(node, "visible", True):
-                logger.debug("  Skipping invisible node")
+    def _is_visible_to_user(self, node: NavigationNode) -> bool:
+        try:
+            # Note how draft status works: there are typically two physical Page objects
+            # for the same logical page: one draft, one published. Which one is returned
+            # depends on the request context, and is handled by Django CMS, by selecting
+            # the appropriate node for that context. Thus, we can assume
```
Comment (pi-sigma) [inline src/open_inwoner/components/templatetags/side_navigation.py]: comment incomplete?

**Review by pi-sigma — APPROVED** [review]
